### PR TITLE
More work on NUOPC cap - DO NOT MERGE YET!

### PR DESCRIPTION
**Pull request by uturuncoglu**
This Pr aims to update NUOPC cap to support configurations under UFS Coastal Model.

### CODE REVIEW

File: src/schism/schism_nuopc_cap.F90
Comment (platipodium) [inline src/schism/schism_nuopc_cap.F90:677]: dclock and mclock should have better descriptive names

File: src/schism/schism_nuopc_cap.F90
Comment (platipodium) [inline src/schism/schism_nuopc_cap.F90:678]: driverCurrTime to keep consistent

File: src/schism/schism_nuopc_cap.F90
Comment (platipodium) [inline src/schism/schism_nuopc_cap.F90:408]: For now, it's ok to use ocn_current_merid, but this is not a universal (or CF-compliant) name.  In the field creation, let's make sure we have CF information present.

File: src/schism/schism_nuopc_cap.F90
Comment (platipodium) [inline src/schism/schism_nuopc_cap.F90:329]: I am not happy with 'dbug'.  That attribute should have a proper name, like 'debug',  or could even be a 'debug level' to allow more fine-grained debugging (later).  Also 'cvalue' is not good, should be logicalValue, e.g.

File: src/schism/schism_esmf_util.F90
Comment (platipodium) [inline src/schism/schism_esmf_util.F90:74]: better local variable name would be isDebug

File: src/schism/schism_esmf_util.F90
Comment (platipodium) [inline src/schism/schism_esmf_util.F90:528]: Let's keep in mind that there are StateUpdate2 etc also ...

File: src/schism/schism_esmf_util.F90
Comment (platipodium) [inline src/schism/schism_esmf_util.F90:544]: 'indx' not well named.  ie and ip are element and node indices, consistent with schism naming,  but indx (and likely ii) should also be semantically named.

File: src/schism/schism_esmf_util.F90
Comment (platipodium) [inline src/schism/schism_esmf_util.F90:582]: Before accepting this change, we have to consider why we would project node vars onto elems. Shouldn't we rather just make sure that the size of farray is matching with the node / elem based vars and then based on meshloc info **not** project?

File: src/schism/schism_esmf_util.F90
Comment (platipodium) [inline src/schism/schism_esmf_util.F90:1428]: We should use the ESMF_Info structure, i.e. InfoSet (but maybe not in this PR, but at a later stage)

**Review by platipodium — CHANGES_REQUESTED** [review]
Please see the individual comments

**Comment by uturuncoglu** [comment]
@platipodium Thanks. I add couple of more things to fix the current. I'll look at your review tomorrow and fix the issues. Thanks again.

**Review by josephzhang8 — COMMENTED** [review]
Thx @uturuncoglu . The changes look fine to me. 

On another thread you seemed to imply you made changes to downtream. Shall we update ufs-coastal and redo RT? THx

@platipodium: can u plz review/approve. Thx

File: src/schism/schism_nuopc_cap.F90
Comment (uturuncoglu) [reply src/schism/schism_nuopc_cap.F90]: @platipodium it is done.

File: src/schism/schism_nuopc_cap.F90
Comment (uturuncoglu) [reply src/schism/schism_nuopc_cap.F90:329]: @platipodium Okay. The issue is that to use RT ufs.configure template file for all the model components, I am using same attribute in the all components. If I change in here, I also need to change in other components to be consistent. Anyway, I'll do it soon and change it to debug level. The value is required since everything in ESMF config side is string.

File: src/schism/schism_esmf_util.F90
Comment (uturuncoglu) [reply src/schism/schism_esmf_util.F90:1485]: @platipodium Yes, it would be better to address it in an other PR.

File: src/schism/schism_esmf_util.F90
Comment (uturuncoglu) [reply src/schism/schism_esmf_util.F90]: done

File: src/schism/schism_esmf_util.F90
Comment (uturuncoglu) [reply src/schism/schism_esmf_util.F90]: done

File: src/schism/schism_nuopc_cap.F90
Comment (uturuncoglu) [reply src/schism/schism_nuopc_cap.F90]: I changed to debug_level and the default is 0 - no debug output such as writing mesh, import and export states in VTK format

File: src/schism/schism_nuopc_cap.F90
Comment (uturuncoglu) [reply src/schism/schism_nuopc_cap.F90]: done

File: src/schism/schism_esmf_util.F90
Comment (uturuncoglu) [reply src/schism/schism_esmf_util.F90]: @platipodium I placed if statements to the function to work differently in node vs element. I did not check the node version yet but I'll do. Please wait for my confirmation that node (with ESMf connector) is also working as expected before merging the PR.

**Comment by uturuncoglu** [comment]
@platipodium I did changes for your review. The last thing is test the datm+schism configuration with ESMF connectors and also using meshloc node, which is default at this point.

**Comment by uturuncoglu** [comment]
@platipodium I checked the datm+schism case with connector and node based implementation. The model seems working but when I checked the import state I could see following weird patten,

Screenshot 2023-11-06 at 4 23 56 PM

As you can see, there is a region with full of zeros in the import state which also aligns with the decomposition elements. I also checked the model native output using VisIt and probably we have similar issue in there too,

Screenshot 2023-11-06 at 4 32 36 PM

This gets more smooth in the model output once the model progress in time. So, we might need to look at more carefully and if you would like to retain the node based approach, this needs to be fixed in the model side. Maybe this is can be fixed if you apply halo update after receiving ESMF fields (maybe you are already doing it, not sure). I could also try with the code version before we introduce the element based approach to eliminate any missing code to retain both approach but I think this was also exist but not exposed. To that end, Have you ever checked import and export states visually before. 

I am not sure but this might not be an issue in CoastalApp since in those configurations uses custom data atmosphere (ATMMESH) in there and it was always providing data with the same resolution of the model and using redist in the connectors. In our case, CDEPS data atmosphere is not in the same resolution with the model and we are using bilinear with NUOPC connectors. Other than this the PR seems good to me. Anyway, let me know how you want to proceed.

**Comment by josephzhang8** [comment]
@uturuncoglu I also recalled something similar (not same) to this, when we were playing with node based partitioning. Halo exchange was part of the problem and Robert O.'s solution seemed to solve the issue eventually. Anyway, my guess this is related now to the halo... thx

**Comment by uturuncoglu** [comment]
@josephzhang8 Yes, it seems like that. Just to clarify, the last run that I mentioned is using node based partitioning along with the NUOPC connectors. So, this case must be same with the original code that supports only node based partitioning. Anyway, I'll look at more to find out any part that needs conditional (element vs. node).

**Review by platipodium — APPROVED** [review]

**Pull request merged**
